library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+
+/*
+ * Generate the String to work on
+ */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+    /*
+     * According to SP800-38F Table 1, the plaintext length for KW
+     * must be between 2 to 2^54-1 semiblocks inclusive.
+     */
+        if( ( in_len < 8 )
+#if SIZE_MAX > UINT_MAX
+                || ( in_len > 0x1FFFFFFFFFFFFF8 )
+#endif
+           )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+
+        memcpy( output, ICV1, KW_SEMIBLOCK_LENGTH );
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+    }
+    else
+    {
+        /*
+         * According to SP800-38F Table 1, the plaintext length for KWP
+         * must be between 1 and 2^32-1 octets inclusive.
+         */
+        if( ( in_len < 1 )
+#if SIZE_MAX > UINT_MAX
+           || ( in_len > 0xFFFFFFFF )
+#endif
+          )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        if( ( in_len % 8 ) != 0 )
+        {
+            padlen = ( 8 - ( in_len % 8 ) );
+        }
+
+        memcpy( output, &ICV2, KW_SEMIBLOCK_LENGTH / 2 );
+        PUT_UINT32_BE( ( in_len & 0xffffffff ), output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        memcpy( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t), input, in_len );
+        memset( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t) + in_len, 0, padlen );
+    }
+    semiblocks = ( ( in_len + padlen ) / KW_SEMIBLOCK_LENGTH ) + 1;
+
+    s = 6 * ( semiblocks - 1 );
+
+    if( ( mode == MBEDTLS_KW_MODE_KWP )
+            && ( in_len <= KW_SEMIBLOCK_LENGTH ) )
+    {
+        memcpy( inbuff, output, 16 );
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx, inbuff, 16, output, &olen);
+        if( ret != 0 )
+            goto cleanup;
+
+        *out_len = olen;
+    }
+    else
+    /*
+     * Do the wrapping function W, as defined in SP 800-38F section 6.1
+     */
+    {
+        if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+            return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+
+        /* Calculate intermediate values */
+        for( t = 1; t <= s; t++ )
+        {
+            memcpy( inbuff, A, KW_SEMIBLOCK_LENGTH );
+            memcpy( inbuff + KW_SEMIBLOCK_LENGTH, R2, KW_SEMIBLOCK_LENGTH );
+
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx, inbuff, 16, outbuff, &olen);
+            if( ret != 0 )
+                goto cleanup;
+
+            memcpy( A, outbuff, KW_SEMIBLOCK_LENGTH );
+            for( i = 0; i < sizeof(t); i++ )
+            {
+                A[i] ^= ((unsigned char*)&t)[sizeof(t)-1-i];
+            }
+
+            /* shift output semiblocks  */
+            for( i = 1; i < semiblocks - 1; i++ )
+                memcpy( output + ( i * KW_SEMIBLOCK_LENGTH ), output + ( ( i + 1 ) * KW_SEMIBLOCK_LENGTH ), KW_SEMIBLOCK_LENGTH );
+
+            /* Set the last semi block as LSB64 of outbuff*/
```

Comment:
Use "semiblock" consistently, per the standard. (Remove space between `semi` and `block`.)